### PR TITLE
:sparkles: add optional ControlPlaneSpec to AWSClusterSpec type

### LINKED ISSUE #873: Possible to provision API with an internal LB ?
/kind feature

Is it possible to add configuration to the cluster network, which would permit switching from "internet-facing" scheme, to "internal" for the control plane API?  From my reading of the docs this would be a new feature request.

I went to add a field to my cluster.yaml during clusterctl create, but failed to grok that it's just a status field in the spec, not anything you can actually change or set I think.

```
status:
  apiEndpoints:
  - host: capi-test2-apiserver-xxx.region-1.elb.amazonaws.com
    port: 443
  providerStatus:
    bastion:
      id: ""
    metadata:
      creationTimestamp: null
    network:
      apiServerElb:
        attributes:
          idleTimeout: 600000000000
        dnsName: capi-test2-apiserver-xxx.region-1.elb.amazonaws.com
        name: capi-test2-apiserver
        scheme: internet-facing
        securityGroupIds:
        - sg-12345
...
```

...so obviously I can't actually set that from the template (or in kubectl edit cluster) what chaos it would be if the cluster API and kube API moved?  There are SSL certs in the cluster providence that depend on the hostname of that ELB to remain stable, I guess that's out.

So I'm assuming this would be a strictly cluster-up decision, like most of cluster-api, and not some kind of public-to-private cluster lifecycle event or vice-versa.  I don't think my organization will ever want to make Kube API public.  I tried taking down the LB manually, but that caused all kinds of havoc with my prometheus metrics and ultimately may have been the cause of death for that cluster.

With a control plane LB in the "internal" scheme mode, we can control access to the API through VPC/VPN and enforce a second factor, and maybe in this way we can safely avoid implementing any more complicated authentication schemes for a while.  Presumably I could restrict access to the load balancer with some security group changes?

I am not expert or maybe not permissioned highly enough to figure that out, after several attempts, it seems like ~perhaps some other networking equipment attached to my VPC is preventing it because of the non-standard port 6443.  I know we have a Cisco that prevents traffic on non-standard ports from leaving the VPC, and the internet-facing LB is outside of the VPC, so~ I think I really need to use an internal load balancer configuration in order to solve this in my environment.

edit: I ~sort of solved the API security, with security groups.~ Editing the security groups in-place was misguided here ~I needed to be sure that the VPC's NAT GW was added to the security group (not my VPN's NAT GW, which is separate), as well as the ELB itself.~

There is a `[CLUSTERNAME]-lb` security group created by the default workflow, which has no rules or attachments.  I added it to my ELB (guessing that my scenario or something like it is probably what this was meant to be used for) and then put a 6443/TCP rule for it in the security group with another one for my NAT GW, and subnet with the cluster nodes in it, and the bastion host for good measure too.

The `schema: internet-facing` ELB is now usable in "private mode", although my traffic must leave the VPC which is less than optimal; I can at least work with this, but it is not a garden path either.  I understand that customizing security groups is not yet in scope for CAPA, maybe internal LB is a good place to start.

Yikes, I understand now that I should not edit the security groups in-place, or I will get a failure to reconcile.  Perhaps when I try this again next time, I will add some more security groups instead.

### PR DESCRIPTION
**Pull request by andrewmyhre**
**What this PR does / why we need it**:
- Allows specification of the control plane load balancer scheme (internet-facing or private). Defaults to internet-facing
- Implements a struct for control plane configuration which can be extended

**Which issue(s) this PR fixes** 
Fixes #873

### CODE REVIEW

File: api/v1alpha2/awscluster_types.go [hunk 1]
```diff
@@ -41,6 +41,16 @@ type AWSClusterSpec struct {
 	// ones added by default.
 	// +optional
 	AdditionalTags Tags `json:"additionalTags,omitempty"`
+
+	// ControlPlaneSpec is optional configuration for customizing control plane behavior
+	// +optional
+	ControlPlaneSpec *AWSControlPlaneSpec `json:controlPlaneSpec,omitempty"`
```
Comment (vincepri) [inline api/v1alpha2/awscluster_types.go]:
```suggestion
	ControlPlaneLoadBalancer *AWSLoadBalancerSpec `json:controlPlaneLoadBalancer,omitempty"`
```

File: api/v1alpha2/awscluster_types.go [hunk 2]
```diff
@@ -41,6 +41,16 @@ type AWSClusterSpec struct {
...
+}
+
+type AWSControlPlaneSpec struct {
+	// LoadBalancerScheme sets the scheme of the control plane load balancer  (defaults to Internet-facing)
+	// +optional
+	ControlPlaneLoadBalancerScheme *ClassicELBScheme `json:"controlPlaneLoadBalancerScheme,omitempty""`
```
Comment (vincepri) [inline api/v1alpha2/awscluster_types.go]:
```suggestion
	Scheme *ClassicELBScheme `json:"scheme,omitempty""`
```

File: api/v1alpha2/awscluster_types.go [hunk 3]
```diff
@@ -41,6 +41,16 @@ type AWSClusterSpec struct {
 	// ones added by default.
 	// +optional
 	AdditionalTags Tags `json:"additionalTags,omitempty"`
+
+	// ControlPlaneSpec is optional configuration for customizing control plane behavior
+	// +optional
+	ControlPlaneSpec *AWSControlPlaneSpec `json:controlPlaneSpec,omitempty"`
+}
+
+type AWSControlPlaneSpec struct {
```
Comment (vincepri) [inline api/v1alpha2/awscluster_types.go]:
```suggestion
type AWSLoadBalancerSpec struct {
```

File: api/v1alpha2/awscluster_types.go (hunk 1 above)
Comment (andrewmyhre) [reply api/v1alpha2/awscluster_types.go]: Addressed

File: api/v1alpha2/awscluster_types.go (hunk 2 above)
Comment (andrewmyhre) [reply api/v1alpha2/awscluster_types.go]: Addressed

File: api/v1alpha2/awscluster_types.go (hunk 3 above)
Comment (andrewmyhre) [reply api/v1alpha2/awscluster_types.go]: Addressed

File: api/v1alpha2/awscluster_types.go [hunk 4]
```diff
@@ -41,6 +41,16 @@ type AWSClusterSpec struct {
 	// ones added by default.
 	// +optional
 	AdditionalTags Tags `json:"additionalTags,omitempty"`
+
+	// ControlPlaneSpec is optional configuration for customizing control plane behavior
+	// +optional
+	ControlPlaneLoadBalancer *AWSLoadBalancerSpec `json:"controlPlaneLoadBalancer,omitempty"`
+}
+
+type AWSLoadBalancerSpec struct {
```
Comment (vincepri) [inline api/v1alpha2/awscluster_types.go]: Add a godoc up here :)

Comment (andrewmyhre) [reply api/v1alpha2/awscluster_types.go]: Added!

File: api/v1alpha2/awscluster_types.go [hunk 5]
```diff
@@ -41,6 +41,17 @@ type AWSClusterSpec struct {
 	// ones added by default.
 	// +optional
 	AdditionalTags Tags `json:"additionalTags,omitempty"`
+
+	// ControlPlaneSpec is optional configuration for customizing control plane behavior
```
Comment (ncdc) [inline api/v1alpha2/awscluster_types.go]: Is this supposed to be ControlPlaneSpec with a nested LoadBalancerSpec, or just ControlPlaneLoadBalancer?

File: pkg/cloud/scope/cluster.go [hunk 6]
```diff
@@ -146,6 +146,19 @@ func (s *ClusterScope) Region() string {
 	return s.AWSCluster.Spec.Region
 }
 
+// ControlPlaneLoadBalancer returns the ControlPlaneSpec
```
Comment (ncdc) [inline pkg/cloud/scope/cluster.go]: ControlPlaneSpec -> something else?

**Pull request merged**